src/Generation/MethodDetails.php

```diff
@@ -131,104 +137,132 @@ private static function maybeCreatePaginated(ServiceDetails $svc, MethodDescript
             ->map(fn ($k) => [$k, $rawFields[$k]])
             ->filter(fn ($x) => $x[1]->isRepeated());
         $resourceByNumber = $resourceCandidates->orderBy(fn ($x) => $x[0])->firstOrNull();
+        if ($isRestOnly) {
+            $resourceListCandidates = $resourceCandidates->filter(fn ($x) => !ProtoHelpers::isMap($catalog, $x[1]));
+            $resourceMapCandidates = $resourceCandidates->filter(fn ($x) => ProtoHelpers::isMap($catalog, $x[1]));
+            // If there are more than one of either, do not generate a paginated method.
+            if (count($resourceListCandidates) > 1 || count($resourceMapCandidates) > 1) {
+                return null;
+            }
+            // A repeated, non-map field takes precedence.
+            $resourceByNumber = $resourceListCandidates->orderBy(fn ($x) => $x[0])->firstOrNull();
+            if (is_null($resourceByNumber)) {
+                $resourceByNumber = $resourceMapCandidates->orderBy(fn ($x) => $x[0])->firstOrNull();
+            }
+        }
         $resourceByPosition = $resourceCandidates->firstOrNull();
         $resources = is_null($resourceByNumber) ? null : $resourceByNumber[1];
         // Valid if it's the first field by position and number, and is not a map.
-        $resourceFieldValid = !is_null($resources) && $resourceByNumber[0] === $resourceByPosition[0] && !ProtoHelpers::isMap($catalog, $resources);
+        $resourceFieldValid = !is_null($resources);
+        // Leverage short-circuting.
+        if ($resourceFieldValid && !$isRestOnly) {
+            $resourceFieldValid &= !ProtoHelpers::isMap($catalog, $resources)
+                && $resourceByNumber[0] === $resourceByPosition[0];
+        }
+
         if (is_null($pageSize) || is_null($pageToken) || is_null($nextPageToken) || is_null($resources)) {
             return null;
+        }
+
+        $isValidPageSize = !$pageSize->isRepeated();
+        if ($isRestOnly) {
+            $isValidPageSize = $pageSize->getType() === GPBType::UINT32 || $pageSize->getType() === GPBType::INT32;
         } else {
-            if ($pageSize->isRepeated() || $pageSize->getType() !== GPBType::INT32) {
-                throw new \Exception("page_size field must be of type int32.");
-            }
-            if ($pageToken->isRepeated() || $pageToken->getType() !== GPBType::STRING) {
-                throw new \Exception("page_token field must be of type string.");
-            }
-            if ($nextPageToken->isRepeated() || $nextPageToken->getType() !== GPBType::STRING) {
-                throw new \Exception("next_page_token field must be of type string.");
+            $isValidPageSize = $pageSize->getType() === GPBType::INT32;
```

Comment:
Potentially I'm reading it wrong (kind of hard to read the diff in github dif page), but it looks like this `else` block has changed the `if` statement it belonged to - it used to belong to `if (is_null($pageSize) || is_null($pageToken) || is_null($nextPageToken) || is_null($resources))` and the rest of the if statements below were nested to it. Now it became the `else` for the `isValidPageSize` and the rest of the if statements are flat (not nested to this else block).  \

The whole sequence may still work as expected, but I wanted to make sure this was intentional. Was it? Even if it was, why was it done this way (basically the diff looks  much bigger than it actually is because of it).

Reply:
This diff is indeed hard to read. The original was something like this:

```
if (is_null($pageSize) || is_null($pageToken) || is_null($nextPageToken) || is_null($resources)) {
    return null;
} else {
    if ($pageSize->isRepeated() || $pageSize->getType() !== GPBType::INT32) {
        throw new \Exception("page_size field must be of type int32.");
    }
   // More logic here.
   return new class // etc.
}
```

However, the `else` block is unnecessary and adds avoidable nesting. So this became:

```
if (is_null($pageSize) || is_null($pageToken) || is_null($nextPageToken) || is_null($resources)) {
    return null;
}

if ($pageSize->isRepeated() || $pageSize->getType() !== GPBType::INT32) {
    throw new \Exception("page_size field must be of type int32.");
}
// More logic here.
return new class // etc.
```

Finally, the `$pageSize` check was altered to factor in REGAPIC-specific criteria:
```
if (is_null($pageSize) || is_null($pageToken) || is_null($nextPageToken) || is_null($resources)) {
    return null;
}

$isValidPageSize = !$pageSize->isRepeated();
if ($isRestOnly) {
$isValidPageSize = $pageSize->getType() === GPBType::UINT32 
    || $pageSize->getType() === GPBType::INT32;
} else {
    $isValidPageSize = $pageSize->getType() === GPBType::INT32;
}
if (!$isValidPageSize) {
  // Throw exception.
}
// etc/
```